src/module/functions/mod.rs

```diff
@@ -418,6 +428,124 @@ impl Module {
+
+                // Run the builder function to produce a new local function, or create a stub that is unreachable
+                let new_local_fn = match fn_builder((&params, &results))
+                    .context("export fn builder failed")?
+                {
+                    Some(f) => f,
+                    None => {
+                        let mut builder = FunctionBuilder::new(&mut self.types, &params, &results);
+                        builder.func_body().unreachable();
+                        builder.local_func(vec![])
+                    }
+                };
+
+                let new_fid = self.funcs.add_local(new_local_fn);
+                let export = self.exports.get_mut(exported_fn.id());
+                export.item = ExportItem::Function(new_fid);
+
+                Ok(new_fid)
+            }
+            // The export didn't exist, or the function isn't the kind we expect
+            _ => bail!("cannot replace function [{fid:?}], it is not an exported function"),
+        }
+    }
+
+    /// Replace a single imported function with the result of the provided builder function.
+    ///
+    /// When called, if `builder` produces a None value, the function in question will be
+    /// replaced with a stub that does nothing (more precisely, a function with an unreachable body).
+    ///
+    /// This function returns the function ID of the *new* function, and
+    /// removes the existing import that has been replaced (the function will become local).
+    pub fn replace_imported_func_by_id<F>(
+        &mut self,
+        fid: FunctionId,
+        fn_builder: F,
+    ) -> Result<FunctionId>
+    where
+        F: FnOnce((&FuncParams, &FuncResults)) -> Result<Option<LocalFunction>>,
+    {
+        // If the function is in the imports, replace it
+        match (self.imports.get_imported_func(fid), self.funcs.get(fid)) {
+            (
+                Some(original_imported_fn),
+                Function {
+                    kind: FunctionKind::Import(ImportedFunction { ty: tid, .. }),
+                    ..
+                },
+            ) => {
+                // Retrieve the params & result types for the imported function
+                let ty = self.types.get(*tid);
+                let (params, results) = (ty.params().to_vec(), ty.results().to_vec());
+
+                // Run the builder function to produce a new local function, or create a stub that is unreachable
+                let new_local_fn = match fn_builder((&params, &results))
+                    .context("import fn builder failed")?
+                {
+                    Some(f) => f,
+                    None => {
+                        let mut builder = FunctionBuilder::new(&mut self.types, &params, &results);
+                        builder.func_body().unreachable();
+                        builder.name("no-op".into());
+                        builder.local_func(vec![])
```

Comment:
I'd prefer we keep stubbing as an implementation detail rather than providing this behaviour by default. We can definitely add an example of stubbing though in the comments if you like.

Reply:
That's reasonable -- will update this and add an example to the documentation (maybe even a doc test?)

- [ ] remove stubbing as default behavior